packages/sources/s3reader/src/transport/csv.ts

```diff
@@ -0,0 +1,129 @@
+import { SubscriptionTransport } from '@chainlink/external-adapter-framework/transports/abstract/subscription'
+import { BaseEndpointTypes, inputParameters } from '../endpoint/csv'
+import { Requester } from '@chainlink/external-adapter-framework/util/requester'
+import { TransportDependencies } from '@chainlink/external-adapter-framework/transports'
+import { EndpointContext } from '@chainlink/external-adapter-framework/adapter'
+import { sleep, AdapterResponse } from '@chainlink/external-adapter-framework/util'
+import { makeLogger } from '@chainlink/external-adapter-framework/util'
+import { S3Client, GetObjectCommand } from '@aws-sdk/client-s3'
+import { parse } from 'csv-parse/sync'
+
+const logger = makeLogger('S3PollerTransport')
+
+export type TransportTypes = BaseEndpointTypes
+type RequestParams = typeof inputParameters.validated
+
+export class S3PollerTransport extends SubscriptionTransport<TransportTypes> {
+  settings!: TransportTypes['Settings']
+  requester!: Requester
+  endpointName!: string
+  s3Client!: S3Client
+
+  async initialize(
+    dependencies: TransportDependencies<TransportTypes>,
+    adapterSettings: TransportTypes['Settings'],
+    endpointName: string,
+    transportName: string,
+  ): Promise<void> {
+    await super.initialize(dependencies, adapterSettings, endpointName, transportName)
+    this.settings = adapterSettings
+    this.requester = dependencies.requester
+    this.endpointName = endpointName
+    this.s3Client = new S3Client({})
+  }
+
+  async backgroundHandler(context: EndpointContext<TransportTypes>, entries: RequestParams[]) {
+    await Promise.all(entries.map(async (param) => this.handleRequest(param)))
+    await sleep(context.adapterSettings.BACKGROUND_EXECUTE_MS)
+  }
+
+  async handleRequest(param: RequestParams) {
+    let response: AdapterResponse<BaseEndpointTypes['Response']>
+    try {
+      response = await this._handleRequest(param)
+    } catch (e) {
+      const errorMessage = e instanceof Error ? e.message : 'Unknown error occurred'
+      logger.error(e, errorMessage)
+      response = {
+        statusCode: 502,
+        errorMessage,
+        timestamps: {
+          providerDataRequestedUnixMs: 0,
+          providerDataReceivedUnixMs: 0,
+          providerIndicatedTimeUnixMs: undefined,
+        },
+      }
+    }
+    await this.responseCache.write(this.name, [{ params: param, response }])
+  }
+
+  async _handleRequest(param: RequestParams): Promise<AdapterResponse<TransportTypes['Response']>> {
+    const { bucket, key, row, column } = param
+    const providerDataRequestedUnixMs = Date.now()
+
+    const csvFileAsStr = await this.getFileFromS3(bucket, key)
+    const answer = this.parseCSV(csvFileAsStr, row, column)
```

Comment:
Could we define a schema definition for the expected csv, and log or throw and error if the csv doesn't match the schema? Or just an assertion that certain columns exist and there are no missing entries if required etc.

Could help catch a class of bugs where the csv is corrupt, or the wrong one is sent.

Reply:
[copied from Alec thread] Added some handling for headers, matchers, and resultField (column). Does this cover it?